### ISSUE TITLE
C-x g is reserved for users, not for major/minor modes

### ISSUE THREAD

**Issue by rnkn** [magit/magit#3517]
`magit-file-mode-map` overwrites the key binding `C-x g`. This key binding is reserved for users and is not for major/minor modes, as specified in the Elisp manual:

> The key sequences bound in a major mode keymap should usually start with `C-c`, followed by a control character, a digit, or `{`, `}`, `<`, `>`, `:` or `;`.  The other punctuation characters are reserved for minor modes, and ordinary letters are reserved for users.

– https://www.gnu.org/software/emacs/manual/html_node/elisp/Major-Mode-Conventions.html

~~~
Mx magit-version RET
Magit 2.13.0, Git 2.18.0, Emacs 26.1, darwin
~~~

**Comment by basil-conto** [comment]
> This key binding is reserved for users

<kbd>C-c</kbd>, not <kbd>C-x</kbd>, followed by an ordinary letter is reserved for users.

**Comment by tarsius** [comment]
That's correct but this gives me a change to mention that I actually plan to violate the conventions and bind <kbd>C-c g</kbd> to `magit-file-popup` in the future.

`magit-file-popup` is such a useful yet still little known command and <kbd>C-c M-g</kbd> such an awful binding for it, I feel violating the conventions is worth it. Of course this will be mentioned in a release announcement and come with instructions on how to undo that gross violation of the conventions.

Having to hold one modifier and press a key, let go of the modifier, press another modifier and a key, let go of the modifier, and finally press another key without a modifier to invoke a popup action is more than I can handle. Since I have locally bound <kbd>C-c g</kbd> to it instead I use this popup much more often.

At the same time I would also like to reconsider changing the binding for `magit-dispatch-popup` for the same reason. I do *not* want to use <kbd>C-x C-g</kbd> because I want to be able to use that to abort the <kbd>C-x</kbd> prefix. Any suggestions?

That's my plans for the time being:

* <kbd>C-x g</kbd> `magit-status`
* <kbd>C-c g</kbd> `magit-file-popup` ("like `magit-status` binding")
* <kbd>C-c M-g</kbd> `magit-file-popup` (backward-compatibility and "like `magit-dispatch-popup` binding")
* <kbd>C-x M-g</kbd> `magit-dispatch-popup` (backward-compatibility)
* <kbd>????</kbd> `magit-dispatch-popup`

**Comment by rnkn** [comment]
Whoops how silly of me. Sorry guys.

Using the `C-x` prefix feels impolite to me for a major mode. I have `C-x g` bound to `bury-buffer`. It's A+.

**Comment by npostavs** [comment]
> Using the `C-x` prefix feels impolite to me for a major mode

`C-x g` isn't exactly a major mode binding though, since it's global (not sure if that makes it more or less polite). Which makes it tricky to find keybindings, since they're supposed to work in buffers with other major modes. The only officially reserved bindings for that would be the minor mode bindings I guess, which would be `C-c <punctuation>`: <kbd>C-c ,</kbd> <kbd>C-c .</kbd> <kbd>C-c [</kbd> <kbd>C-c ]</kbd> <kbd>C-c -</kbd> <kbd>C-c =</kbd> (not counting shitfted punctuation, and `{}<>:;` which are reserved for major modes). Of course, since there are so few, these are liable to be shadowed by other minor modes...

**Comment by rnkn** [comment]
Ah yes true, it's technically `magit-file-mode`, a minor mode.

**Comment by phil-s** [comment]
tarsius, please don't do that.  The only sane way to maintain conventions like the `C-c LETTER` reservation is to adhere to them absolutely.  Any number of library authors could decide that there was some reasonable justification for breaking the rules and using one of those key sequences, and they might all have good reasons, but the end result would be messy.

I don't disagree that `C-x M-g` is awkward, and I'm all for *recommending* that users define a better binding such as `C-c g`, but you absolutely shouldn't set that binding yourself.

**Comment by tarsius** [comment]
But think of the kids! (Always a good argument to do anything.)

More specifically non-Emacs users who might be incited by all the "I use Magit for version control, even when working on Java code in *IDE". For someone who never used Emacs having to add `(define-key magit-mode-map "\C-cg\" 'magit-file-popup)` could be a major turn off (or just reading the documentation about why and how to do it). While doing the opposite is absolutely trivial for someone who already used Emacs before.

But yeah, I probably won't do it.

**Comment by rnkn** [comment]
So how do I fix this so I get my `C-x g` back? I've tried setting and saving `global-magit-file-mode` to off, but Magit appears to override this outside Customize...

**Comment by tarsius** [comment]
I have just disabled this mode using the Custom interface and it worked.

I also tried doing so using `(setq global-magit-file-mode nil)` *before* loading `magit`, which also worked. Doing so *after* loading `magit` (`magit-file` actually) would not work, but `(global-magit-file-mode -1)` should.

**Comment by tarsius** [comment]
Also see #3522.

**Comment by rnkn** [comment]
Hmm, customizing `global-magit-file-mode` and saving only works for the current session. Upon reloading Emacs, Magit will override the saved user option and again set `global-magit-file-mode -> t`.

**Comment by rnkn** [comment]
That's not it I'm afraid. I'm loading `custom-set-variables` before anything else: https://github.com/rnkn/.emacs.d/blob/master/init.el#L1-L9

**Comment by fice-t** [comment]
@tarsius 
>`magit-file-popup` is such a useful yet still little known command and <kbd>C-c M-g</kbd> such an awful binding for it, I feel violating the conventions is worth it.

What about binding `magit-file-popup` to <kbd> C-c C-f</kbd>? It's not reserved, it's easy to hit, and I believe it's more mnemonic.

**Comment by phil-s** [comment]
> What about binding `magit-file-popup` to <kbd>C-c C-f</kbd>? It's not reserved

`(elisp)Key Binding Conventions` says: "Sequences consisting of ‘C-c’ followed by a control character or a digit are reserved for major modes."

`magit-file-mode` is a minor mode, so that's not compliant either.

**Comment by phil-s** [comment]
It isn't really any different to running `magit-status` though.  There is no default key-binding for that -- users have to add one, and `(magit)Getting Started` makes an appropriate recommendation.

I suggest that the same node should recommend that users bind `magit-file-popup` to `C-c g`, and then both the important commands are dealt with.

**Comment by fice-t** [comment]
Whoops, I completely forgot about that reservation :sweat: . I guess I've been away from Emacs development for too long... 

> I suggest that the same node should recommend that users bind magit-file-popup to C-c g, and then both the important commands are dealt with.

Agreed. On a related note, I've opened #3663 which would help with the visibility of `magit-file-popup`.
